g2p/mappings/tokenizer.py

```diff
@@ -0,0 +1,204 @@
+                if ipa_successors:
+                    # in_lang has an ipa successor, tokenize using it
+                    # there currently are no langs with more than 1 IPA successor, but to
+                    # be future-proof we'll arbitrary take the first if there are more.
+                    out_lang = ipa_successors[0]
+                else:
+                    # There is no direct IPA successor, look for a two-hop path to -ipa
+                    for x in successors:
+                        ipa_successors_two_hops = [
+                            y for y in LANGS_NETWORK.successors(x) if y.endswith("-ipa")
+                        ]
+                        # LOGGER.warning(pprint.pformat([in_lang, x, "->", [ipa_successors_two_hops]]))
+                        if ipa_successors_two_hops:
+                            out_lang = [x, ipa_successors_two_hops[0]]
+                        break
+                    # There is no two-hop IPA successor, use the first direct successor
+                    if out_lang is None and successors:
+                        out_lang = successors[0]
+            #LOGGER.warning(f"Tokenizer for {in_lang} is {out_lang}.")
+            if out_lang is None:
+                # Default tokenizer:
+                self.tokenizers[tokenizer_key] = self.tokenizers[None]
+            elif isinstance(out_lang, list):
+                # Build a two-hop tokenizer
+                assert len(out_lang) == 2
+                try:
+                    mapping1 = Mapping(in_lang=in_lang, out_lang=out_lang[0])
+                    mapping2 = Mapping(in_lang=out_lang[0], out_lang=out_lang[1])
+                    self.tokenizers[tokenizer_key] = TwoHopTokenizer(
+                        (mapping1, mapping2)
+                    )
+                except MappingMissing:
+                    self.tokenizers[tokenizer_key] = self.tokenizers[None]
+                    LOGGER.warning(
+                        f"missing mapping yet we looked for mappings in graph for {in_lang}-{out_lang}."
+                    )
+            else:
+                # Build a one-hop tokenizer
+                try:
+                    mapping = Mapping(in_lang=in_lang, out_lang=out_lang)
+                    self.tokenizers[tokenizer_key] = Tokenizer(mapping)
+                except MappingMissing:
+                    self.tokenizers[tokenizer_key] = self.tokenizers[None]
+                    LOGGER.warning(
+                        f"Cannot find mapping from '{in_lang}' to '{out_lang}'. Using default tokenizer instead!!!"
+                    )
+
+            # Hack for Tlingit using dot as a letter when non word-final
+            if in_lang == "tli":
+                self.tokenizers[tokenizer_key].dot_is_letter = True
+
+        return self.tokenizers.get(tokenizer_key)
+
+
+_TOKENIZER_LIBRARY = TokenizerLibrary()
+
+
+def get_tokenizer(in_lang=None, out_lang=None, tok_path=None):
```

Comment:
Nice. This is clear, and a very reasonable strategy I think.

Reply:
Thanks, I'm glad you like it. It took me a while to get something I was happy with!